Register custom topics via schemas-tools in kafka-init

It might be nice to have a way of easily registering custom topics via the `kafka-init` mechanism. Part of that is already there, with the `etc/schema/` volume that gets merged. That way the topic can be created. However then the schemas-tools are missing the respective class for registering the topic with the schema-registry (see log).
What I did is building my own schemas-tools that include the class and loading those instead of the github release in the `Dockerfile`. But maybe this can be done in a nicer way? Unless I missed something that already implements that.

```
kafka-init_1                   | 2017-11-13T20:07:16.014342553Z [main] ERROR org.radarcns.schema.util.Utils - Failed to apply function, returning empty.
kafka-init_1                   | 2017-11-13T20:07:16.014729136Z java.lang.IllegalStateException: Topic android_eeg_sync_pulse schema cannot be instantiated
kafka-init_1                   | 2017-11-13T20:07:16.014874882Z 	at org.radarcns.config.AvroTopicConfig.parseAvroTopic(AvroTopicConfig.java:46)
kafka-init_1                   | 2017-11-13T20:07:16.014894100Z 	at org.radarcns.schema.specification.DataTopic.getTopics(DataTopic.java:61)
kafka-init_1                   | 2017-11-13T20:07:16.014904177Z 	at org.radarcns.schema.util.Utils.lambda$applyOrEmpty$0(Utils.java:112)
kafka-init_1                   | 2017-11-13T20:07:16.014927819Z 	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267)
kafka-init_1                   | 2017-11-13T20:07:16.014939647Z 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
kafka-init_1                   | 2017-11-13T20:07:16.014955428Z 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
kafka-init_1                   | 2017-11-13T20:07:16.014979859Z 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
kafka-init_1                   | 2017-11-13T20:07:16.014997257Z 	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
kafka-init_1                   | 2017-11-13T20:07:16.015008452Z 	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
kafka-init_1                   | 2017-11-13T20:07:16.015019391Z 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
kafka-init_1                   | 2017-11-13T20:07:16.015033303Z 	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
kafka-init_1                   | 2017-11-13T20:07:16.015059088Z 	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
kafka-init_1                   | 2017-11-13T20:07:16.015070639Z 	at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1620)
kafka-init_1                   | 2017-11-13T20:07:16.015086843Z 	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
kafka-init_1                   | 2017-11-13T20:07:16.015101537Z 	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
kafka-init_1                   | 2017-11-13T20:07:16.015115079Z 	at java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
kafka-init_1                   | 2017-11-13T20:07:16.015155521Z 	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
kafka-init_1                   | 2017-11-13T20:07:16.015190318Z 	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
kafka-init_1                   | 2017-11-13T20:07:16.015196351Z 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
kafka-init_1                   | 2017-11-13T20:07:16.015205266Z 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
kafka-init_1                   | 2017-11-13T20:07:16.015215939Z 	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
kafka-init_1                   | 2017-11-13T20:07:16.015242588Z 	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
kafka-init_1                   | 2017-11-13T20:07:16.015255832Z 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
kafka-init_1                   | 2017-11-13T20:07:16.015270802Z 	at java.util.stream.ReferencePipeline.allMatch(ReferencePipeline.java:454)
kafka-init_1                   | 2017-11-13T20:07:16.015285701Z 	at org.radarcns.schema.registration.SchemaRegistry.registerSchemas(SchemaRegistry.java:74)
kafka-init_1                   | 2017-11-13T20:07:16.015301602Z 	at org.radarcns.schema.registration.SchemaRegistry$RegisterCommand.execute(SchemaRegistry.java:167)
kafka-init_1                   | 2017-11-13T20:07:16.015320640Z 	at org.radarcns.schema.CommandLineApp.main(CommandLineApp.java:191)
kafka-init_1                   | 2017-11-13T20:07:16.015454724Z Caused by: java.lang.IllegalArgumentException: Topic android_eeg_sync_pulse schema cannot be instantiated
kafka-init_1                   | 2017-11-13T20:07:16.015478518Z 	at org.radarcns.topic.AvroTopic.parse(AvroTopic.java:135)
kafka-init_1                   | 2017-11-13T20:07:16.015490273Z 	at org.radarcns.config.AvroTopicConfig.parseAvroTopic(AvroTopicConfig.java:44)
kafka-init_1                   | 2017-11-13T20:07:16.015513243Z 	... 26 more
kafka-init_1                   | 2017-11-13T20:07:16.015623872Z Caused by: java.lang.ClassNotFoundException: org.radarcns.passive.eegsync.EegSyncPulse
kafka-init_1                   | 2017-11-13T20:07:16.015651514Z 	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
kafka-init_1                   | 2017-11-13T20:07:16.015673188Z 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
kafka-init_1                   | 2017-11-13T20:07:16.015698230Z 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
kafka-init_1                   | 2017-11-13T20:07:16.015730638Z 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
kafka-init_1                   | 2017-11-13T20:07:16.015744674Z 	at java.lang.Class.forName0(Native Method)
kafka-init_1                   | 2017-11-13T20:07:16.015760521Z 	at java.lang.Class.forName(Class.java:264)
kafka-init_1                   | 2017-11-13T20:07:16.015770986Z 	at org.radarcns.topic.AvroTopic.parse(AvroTopic.java:124)
kafka-init_1                   | 2017-11-13T20:07:16.015779249Z 	... 27 more
```

--- a/dcompose-stack/radar-cp-hadoop-stack/hdfs_restructure.sh
+++ b/dcompose-stack/radar-cp-hadoop-stack/hdfs_restructure.sh
@@ -8,7 +8,7 @@ fi
 . ./util.sh
 
 # HDFS restructure version
-JAR_VERSION=0.2.1
+JAR_VERSION=0.3.1
 # HDFS restructure JAR
 JAR="restructurehdfs-all-${JAR_VERSION}.jar"
 

--- a/dcompose-stack/radar-cp-hadoop-stack/install-radar-stack.sh
+++ b/dcompose-stack/radar-cp-hadoop-stack/install-radar-stack.sh
@@ -10,6 +10,8 @@ else
   sudo-linux mkdir -m 0700 ./output
 fi
 
+check_config_present .env etc/env.template
+
 . ./.env
 
 # Check provided directories and configurations

--- a/dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/Dockerfile
+++ b/dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/Dockerfile
@@ -8,17 +8,30 @@ ENV KAFKA_NUM_REPLICATION=2
 ENV KAFKA_NUM_BROKERS=3
 ENV KAFKA_ZOOKEEPER_CONNECT=zookeeper-1:2181
 
-RUN apk add --no-cache rsync tar bash wget
+RUN apk add --no-cache \
+  bash \
+  curl \
+  rsync \
+  tar
 
-RUN mkdir -p /schema/merged
+RUN mkdir -p /schema/merged /schema/java/src /schema/java/classes /usr/share/java
 
 WORKDIR /schema
 
-ADD https://github.com/RADAR-CNS/RADAR-Schemas/archive/v0.2.tar.gz /schema/RADAR-Schemas-0.2.tar.gz
-RUN tar xzf RADAR-Schemas-0.2.tar.gz && mv RADAR-Schemas-0.2 original && rm RADAR-Schemas-0.2.tar.gz
+ENV JQ_VERSION=1.5
+RUN curl -L#o /usr/bin/jq https://github.com/stedolan/jq/releases/download/jq-${JQ_VERSION}/jq-linux64 \
+    && chmod +x /usr/bin/jq
+RUN curl -#o /usr/share/java/avro-tools.jar \
+ "$(curl -s http://www.apache.org/dyn/closer.cgi/avro/\?as_json \
+  | jq --raw-output ".preferred")avro/avro-1.8.2/java/avro-tools-1.8.2.jar"
 
-ADD https://github.com/RADAR-CNS/RADAR-Schemas/releases/download/v0.2/radar-schemas-tools-0.2.tar.gz /schema/radar-schemas-tools-0.2.tar.gz
-RUN tar xzf radar-schemas-tools-0.2.tar.gz --strip-components=1 -C /usr && rm radar-schemas-tools-0.2.tar.gz
+ENV RADAR_SCHEMAS_VERSION=0.2
+RUN curl -#L https://github.com/RADAR-CNS/RADAR-Schemas/releases/download/v${RADAR_SCHEMAS_VERSION}/radar-schemas-tools-${RADAR_SCHEMAS_VERSION}.tar.gz \
+  | tar xz -C /usr --strip-components 1
+
+RUN mkdir original \
+  && curl -#L https://github.com/RADAR-CNS/RADAR-Schemas/archive/v${RADAR_SCHEMAS_VERSION}.tar.gz \
+  | tar xz -C original --strip-components 1
 
 VOLUME /schema/conf
 

--- a/dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/init.sh
+++ b/dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/init.sh
@@ -1,6 +1,18 @@
 #!/bin/bash
 
+set -e
+
 rsync -a /schema/original/commons /schema/original/specifications /schema/merged
 rsync -a /schema/conf/ /schema/merged
+
+# Compiling updated schemas
+echo "Compiling schemas..."
+# Separate enums so that they can be referenced in later files
+IFS=$'\n' read -r -a enums <<< $(find merged/commons -name "*.avsc" -exec grep -q '^  "type": "enum"' "{}" \; -print)
+IFS=$'\n' read -r -a notenums <<< $(find merged/commons -name "*.avsc" -exec grep -qv '^  "type": "enum"' "{}" \; -print)
+java -jar /usr/share/java/avro-tools.jar compile -string schema ${enums[@]} ${notenums[@]} java/src 2>/dev/null
+find java/src -name "*.java" -print0 | xargs -0 javac -cp /usr/lib/*:java/classes -d java/classes -sourcepath java/src 
+# Update the radar schemas so the tools find the new classes in classpath
+jar uf /usr/lib/radar-schemas-commons-${RADAR_SCHEMAS_VERSION}.jar -C java/classes .
 
 exec "$@"

--- a/dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/topic_init.sh
+++ b/dcompose-stack/radar-cp-hadoop-stack/kafka-radarinit/topic_init.sh
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
 
 # Create topics
 echo "Creating RADAR-CNS topics..."
@@ -17,7 +17,7 @@ tries=10
 timeout=1
 max_timeout=32
 while true; do
-  if wget --spider -q "${KAFKA_SCHEMA_REGISTRY}" 2>/dev/null; then
+  if curl -Ifs "${KAFKA_SCHEMA_REGISTRY}" > /dev/null; then
     break;
   fi
   tries=$((tries - 1))
@@ -27,7 +27,9 @@ while true; do
   fi
   echo "Failed to reach schema registry. Retrying in ${timeout} seconds."
   sleep $timeout
-  timeout=$((timeout * 2 < max_timeout ? timeout * 2 : max_timeout))
+  if [ $timeout -lt $max_timeout ]; then
+    timeout=$((timeout * 2))
+  fi
 done
 
 if ! radar-schemas-tools register "${KAFKA_SCHEMA_REGISTRY}" merged; then

--- a/dcompose-stack/radar-cp-hadoop-stack/util.sh
+++ b/dcompose-stack/radar-cp-hadoop-stack/util.sh
@@ -97,9 +97,8 @@ check_config_present() {
     fi
     exit 1
   elif [ "$1" -ot "${template}" ]; then
-    echo "Configuration file ${1} is older than its template"
-    echo "${template}. Please edit ${1}"
-    echo "to ensure it matches the template, remove it or run touch on it."
+    echo "Configuration file ${1} is older than its template ${template}."
+    echo "Please edit ${1} to ensure it matches the template or run touch on it."
     exit 1
   fi
 }

